eth_keys/libsecp256k1.nim

```diff
@@ -0,0 +1,305 @@
+#
+# Nim Ethereum Keys (nim-eth-keys)
+# Copyright (c) 2018 Status Research & Development GmbH
+# Licensed under either of
+# - Apache License, version 2.0, (LICENSE-APACHEv2)
+# - MIT license (LICENSE-MIT)
+#
+
+## This is libsecp256k1 backend.
+
+import secp256k1, nimcrypto/sysrand, nimcrypto/utils
+
+const
+  KeyLength* = 256 div 8
+  InvalidPrivateKey = "Invalid private key!"
+  InvalidPublicKey = "Invalid public key!"
+  InvalidSignature = "Invalid signature!"
+  VerificationFailed = "Signature verification has been failed!"
+  EmptyMessageError = "Data/Message could not be empty!"
+  MessageSizeError = "Size of message to sign must be KeyLength bytes!"
+
+type
+  PublicKey* = secp256k1_pubkey
+    ## Representation of public key
+
+  PrivateKey* = object
+    ## Representation of secret key
+    data*: array[KeyLength, byte]
+
+  SharedSecret* = object
+    ## Representation of ECDH shared secret
+    data*: array[KeyLength, byte]
+
+  Nonce* = object
+    ## Representation of nonce
+    data*: array[KeyLength, byte]
+
+  RawPublicKey* = object {.packed.}
+    ## Representation of serialized public key
+    header*: byte
+    data*: array[KeyLength * 2, byte]
+
+  KeyPair* = object
+    ## Representation of private/public keys pair
+    seckey*: PrivateKey
+    pubkey*: PublicKey
+
+  Signature* = secp256k1_ecdsa_recoverable_signature
+    ## Representation of signature
+
+  RawSignature* = object {.packed.}
+    ## Representation of serialized signature
+    data*: array[KeyLength * 2 + 1, byte]
+
+  Secp256k1Exception* = object of Exception
+    ## Exceptions generated by `libsecp256k1`
+
+  EthKeysContext = ref object of RootRef
+    context: ptr secp256k1_context
+    error: string
+
+var ekContext {.threadvar.}: EthKeysContext
+  ## Thread local variable which holds current context
+
+##
+## Private procedures interface
+##
+
+proc illegalCallback(message: cstring; data: pointer) {.cdecl.} =
+  let ctx = cast[EthKeysContext](data)
+  ctx.error = $message
+
+proc errorCallback(message: cstring, data: pointer) {.cdecl.} =
+  let ctx = cast[EthKeysContext](data)
+  ctx.error = $message
+
+proc newEthKeysContext(): EthKeysContext =
+  ## Create new `EthKeysContext`.
+  result = new EthKeysContext
```

Comment:
EthKeysContext needs a finalizer that will call secp256k1_context_destroy

Reply:
Here is an example of wrapping a C library and use Nim finalizers to call a destroy/free procedure when the ref objects goes out of scope: https://github.com/mratsim/Arraymancer/blob/master/src/tensor/backend/cuda.nim#L23-L40

Reply:
I dont know in what use-case you will need this feature, but ok, i will add it.

Reply:
It's to make sure that eth_keys properly deallocates the memory (context) it allocated.

Reply:
I understand, but when you deallocate this memory context? I think you will deallocate it only on program exit... And OS will deallocate it for you.

Reply:
With finalizer it's deallocated as soon as there is no references to it. If a program runs for a long time but only use secp256k1 in few places that avoids keeping a secp context open.